File: pkmnapi-api/src/requests/type_names.rs
use rocket_okapi::JsonSchema;
use serde::Deserialize;

use crate::requests::base::BaseRequest;

pub type TypeNameRequest = BaseRequest<TypeNameRequestType, TypeNameRequestAttributes>;

impl TypeNameRequest {
    pub fn get_name(&self) -> &String {
        &self.data.attributes.name
    }
}

#[derive(Debug, Deserialize, JsonSchema)]
#[allow(non_camel_case_types)]
pub enum TypeNameRequestType {
    type_names,
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct TypeNameRequestAttributes {
    pub name: String,
}

File: clients/js/src/requests/type_names.ts
import { BaseRequest } from './base';

export type TypeNameRequest = BaseRequest<TypeNameRequestType, TypeNameRequestAttributes>;

enum TypeNameRequestType {
    type_names = 'type_names',
}

export interface TypeNameRequestAttributes {
    name: string;
}

export const TypeNameRequestDefault = (attributes: TypeNameRequestAttributes): TypeNameRequest => {
    return {
        data: {
            type: TypeNameRequestType.type_names,
            attributes,
        },
    };
};
